### PR TITLE
Redis MRR Support

### PR DESCRIPTION
**Pull request by himanikh**
Changes to support Multi read replica feature on Cloud Memorystor redis

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know). - Unable to run these tests target testacc unavailable
- [ X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
redis: Added Multi read replica field `replicaCount `, `nodes`,  `readEndpoint`, `readEndpointPort`, `readReplicasMode` in `google_redis_instance `

```

### CODE REVIEW

File: mmv1/products/redis/terraform.yaml [hunk 1]
```diff
@@ -53,14 +53,22 @@ overrides: !ruby/object:Overrides::ResourceOverrides
...
+          network_name: 'BootstrapSharedTestNetwork(t, "redis-mrr")'
+        min_version: beta
     properties:
       alternativeLocationId: !ruby/object:Overrides::Terraform::PropertyOverride
         default_from_api: true
       authorizedNetwork: !ruby/object:Overrides::Terraform::PropertyOverride
         default_from_api: true
         custom_expand: 'templates/terraform/custom_expand/redis_instance_authorized_network.erb'
         diff_suppress_func: 'compareSelfLinkOrResourceName'
-      locationId: !ruby/object:Overrides::Terraform::PropertyOverride
```
Comment (melinath) [inline mmv1/products/redis/terraform.yaml:63]: This seems like an extraneous change

**Review by melinath — CHANGES_REQUESTED** [review]
Please add a release note as described in the opening comment's template

File: mmv1/products/redis/terraform.yaml (hunk 1 above)
Comment (himanikh) [reply mmv1/products/redis/terraform.yaml]: added back

File: mmv1/templates/terraform/examples/redis_instance_mrr.tf.erb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  location_id             = "us-central1-a"
+  alternative_location_id = "us-central1-f"
+
+  authorized_network = data.google_compute_network.redis-network.id
+
+  redis_version     = "REDIS_6_X"
+  display_name      = "Terraform Test Instance"
+  reserved_ip_range = "192.168.0.0/29"
+  replica_count     = 5
+  read_replicas_mode = "READ_REPLICAS_ENABLED"
```
Comment (melinath) [inline mmv1/templates/terraform/examples/redis_instance_mrr.tf.erb:16]: would it be possible to also add values for readEndpoint and readEndpointPort here?

**Review by melinath — CHANGES_REQUESTED** [review]
In addition to the changes below, please update your release note per the guidelines and sign the CLA as described here: https://github.com/GoogleCloudPlatform/magic-modules/pull/5449/checks?check_run_id=4196015153

File: mmv1/products/redis/api.yaml [hunk 3]
```diff
@@ -229,3 +229,56 @@ objects:
               output: true
               description: |
                 Sha1 Fingerprint of the certificate.
+      - !ruby/object:Api::Type::Integer
+        name: replicaCount
+        description: |
+          Optional. The number of replica nodes. Valid range for standard tier is [1-5]
+          and defaults to 2. Valid value for basic tier is 0 and defaults to 0.
+        min_version: beta
```
[the remark this answers is not in the document]
Comment (melinath) [reply mmv1/products/redis/api.yaml:237]: Per discussion: the feature that these fields support is in public preview, even though the fields are also available from the GA endpoint.

File: mmv1/templates/terraform/examples/redis_instance_mrr.tf.erb (hunk 2 above)
Comment (melinath) [reply mmv1/templates/terraform/examples/redis_instance_mrr.tf.erb:16]: per discussion, those are read-only.

File: mmv1/products/redis/api.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (himanikh) [reply mmv1/products/redis/api.yaml:237]: As discussed the feature is still in public preview and thus would keep the min_version: beta

File: mmv1/products/redis/api.yaml [hunk 4]
```diff
@@ -229,3 +229,56 @@ objects:
...
+        description: |
+          The port number of the exposed readonly redis endpoint. Standard tier only. 
+          Write requests should target 'port'.
+        output: true
+        min_version: beta
+      - !ruby/object:Api::Type::Enum
+        name: readReplicasMode
+        description: |
+          Optional. Read replica mode.
+          If not set, Memorystore Redis backend will pick the mode based on other fields in the request.
```
[the remark this answers is not in the document]
Comment (himanikh) [reply mmv1/products/redis/api.yaml]: Updated the description.

**Comment by melinath** [comment]
TestAccRedisInstance_redisInstanceMrrExample  is failing with the errors:

```
Error: Error creating Instance: googleapi: Error 400: Invalid IP range: 192.168.0.0/29. Require a minimum of /28 block size when read replica is enabled.
com.google.apps.framework.request.StatusException: <eye3 title='INVALID_ARGUMENT'/> generic::INVALID_ARGUMENT: Invalid IP range: 192.168.0.0/29. Require a minimum of /28 block size when read replica is enabled.
Details:
[
{
"@type": "type.googleapis.com/google.rpc.BadRequest",
"fieldViolations": [
{
"description": "Invalid value: 192.168.0.0/29",
"field": "instance.reserved_ip_range"
}
]
}
]
on terraform_plugin_test.tf line 2, in resource "google_redis_instance" "cache":
2: resource "google_redis_instance" "cache" {
```

Probably the test needs to be updated to have a valid value.

---

Additionally, TestAccRedisInstance_redisInstanceFullExample is failing with the error:

```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
~ update in-place
Terraform will perform the following actions:
# google_redis_instance.cache will be updated in-place
~ resource "google_redis_instance" "cache" {
id                       = "projects/ci-test-project-188019/locations/us-central1/instances/tf-test-ha-memory-cache7zfrzip0gk"
name                     = "tf-test-ha-memory-cache7zfrzip0gk"
- replica_count            = 1 -> null
# (23 unchanged attributes hidden)
}
Plan: 0 to add, 1 to change, 0 to destroy.
```

This probably means that the API server sets a default value of 1 if the replica_count isn't set in the request. If that field is likely to always default to 1, then you could resolve this by adding `default_value: 1` to replica_count in api.yaml - otherwise, you could add `default_from_api: true` as an override in terraform.yaml (search the codebase for examples)

**Review by melinath — CHANGES_REQUESTED** [review]
see above comment - forgot to leave it as a change request

File: mmv1/products/redis/api.yaml [hunk 5]
```diff
@@ -229,3 +229,58 @@ objects:
...
+        name: readEndpointPort
+        description: |
+          Output only. The port number of the exposed readonly redis endpoint. Standard tier only. 
+          Write requests should target 'port'.
+        output: true
+        min_version: beta
+      - !ruby/object:Api::Type::Enum
+        name: readReplicasMode
+        description: |
+          Optional. Read replica mode. Can only be specified when trying to create the instance.
```
Comment (melinath) [inline mmv1/products/redis/api.yaml:276]: readReplicasMode should also have `input: true` set here if it can only be set on create.

Comment (himanikh) [reply mmv1/products/redis/api.yaml:276]: done

**Review by melinath — APPROVED** [review]
lgtm - remaining failures are unrelated

**Pull request merged**